File: lake/src/lake/ingest/rss.py
import json
import uuid
from datetime import datetime, timezone

import feedparser
import httpx
import pyarrow as pa

from datalake import write_bronze, DatalakeConfig, BRONZE_SCHEMA

FEEDS = {
    "guardian_au_business": "https://www.theguardian.com/au/business/rss",
    "rba_media_releases": "https://www.rba.gov.au/rss/rss-cb-media-releases.xml",
    "rba_speeches": "https://www.rba.gov.au/rss/rss-cb-speeches.xml",
}


def fetch_feeds(feeds: dict[str, str] | None = None, timeout: int = 10) -> list[dict]:
    feeds = feeds or FEEDS
    articles = []
    for source_id, url in feeds.items():
        try:
            resp = httpx.get(url, follow_redirects=True, timeout=timeout)
            resp.raise_for_status()
            feed = feedparser.parse(resp.text)
        except Exception as e:
            print(f"Warning: failed to fetch {source_id} ({url}): {e}")
            continue

        for entry in feed.entries:
            published = ""
            if hasattr(entry, "published_parsed") and entry.published_parsed:
                published = datetime(
                    *entry.published_parsed[:6], tzinfo=timezone.utc
                ).isoformat()
            elif hasattr(entry, "updated_parsed") and entry.updated_parsed:
                published = datetime(
                    *entry.updated_parsed[:6], tzinfo=timezone.utc
                ).isoformat()

            articles.append(
                {
                    "title": getattr(entry, "title", ""),
                    "url": getattr(entry, "link", ""),
                    "published_at": published,
                    "source": source_id,
                    "summary": getattr(entry, "summary", ""),
                }
            )
    return articles


def build_bronze_table(articles: list[dict], batch_id: str | None = None) -> pa.Table:
    batch_id = batch_id or str(uuid.uuid4())
    now = datetime.now(timezone.utc)
    return pa.table(
        {
            "_source": [f"rss.{a['source']}" for a in articles],
            "_ingested_at": pa.array(
                [now] * len(articles), type=pa.timestamp("us", tz="UTC")
            ),
            "_raw_payload": [json.dumps(a) for a in articles],
            "_batch_id": [batch_id] * len(articles),
        },
        schema=BRONZE_SCHEMA,
    )


def ingest(config: DatalakeConfig | None = None) -> str:
    articles = fetch_feeds()
    if not articles:
        print("No articles fetched from any feed")
        return ""
    table = build_bronze_table(articles)
    path = write_bronze(table, "rss", "news.parquet", config=config)
    print(f"Wrote {len(articles)} articles to {path}")
    return path

File: lake/internal/ingest/rss.go
package ingest

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"time"

	"github.com/google/uuid"
	"github.com/mmcdole/gofeed"
	"github.com/solanyn/mono/lake/internal/metrics"
	"github.com/solanyn/mono/lake/internal/storage"
)

var rssFeeds = map[string]string{
	"guardian_au_business": "https://www.theguardian.com/au/business/rss",
	"rba_media_releases":   "https://www.rba.gov.au/rss/rss-cb-media-releases.xml",
	"rba_speeches":         "https://www.rba.gov.au/rss/rss-cb-speeches.xml",
}

func IngestRSS(ctx context.Context, s3 *storage.Client) error {
	start := time.Now()
	source := "rss"

	fp := gofeed.NewParser()
	var rows []map[string]interface{}

	for sourceID, url := range rssFeeds {
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
		if err != nil {
			log.Printf("rss: skip %s: %v", sourceID, err)
			continue
		}
		resp, err := http.DefaultClient.Do(req)
		if err != nil {
			log.Printf("rss: skip %s: %v", sourceID, err)
			continue
		}
		body, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			continue
		}

		feed, err := fp.ParseString(string(body))
		if err != nil {
			log.Printf("rss: parse %s: %v", sourceID, err)
			continue
		}

		for _, item := range feed.Items {
			published := ""
			if item.PublishedParsed != nil {
				published = item.PublishedParsed.UTC().Format(time.RFC3339)
			} else if item.UpdatedParsed != nil {
				published = item.UpdatedParsed.UTC().Format(time.RFC3339)
			}
			rows = append(rows, map[string]interface{}{
				"title":        item.Title,
				"url":          item.Link,
				"published_at": published,
				"source":       sourceID,
				"summary":      item.Description,
			})
		}
	}

	if len(rows) == 0 {
		log.Println("rss: no articles fetched")
		return nil
	}

	batchID := uuid.New().String()
	data, err := storage.WriteBronze(rows, "rss", batchID)
	if err != nil {
		metrics.IngestErrors.WithLabelValues(source).Inc()
		return fmt.Errorf("write bronze: %w", err)
	}

	key, err := s3.PutParquet(ctx, "bronze", "rss", "news.parquet", data)
	if err != nil {
		metrics.IngestErrors.WithLabelValues(source).Inc()
		return fmt.Errorf("put s3: %w", err)
	}

	metrics.IngestTotal.WithLabelValues(source).Inc()
	metrics.IngestDuration.WithLabelValues(source).Observe(time.Since(start).Seconds())
	metrics.LastIngestTimestamp.WithLabelValues(source).SetToCurrentTime()
	log.Printf("rss: wrote %d articles to %s", len(rows), key)
	return nil
}

func init() {
	_ = json.Marshal
}
